### PR TITLE
Fix ghost files and improve error handling

### PR DESCRIPTION
**Pull request by Snorch**
Port patches from Virtuozzo criu:

files-reg: temporary remount writable the mount we do unlink on
ghost/mount: allocate remounted_rw in shmem to get info from other processes
files-reg: fix error handling of rm_parent_dirs
files-reg: fix error handling in open_path

Should fix the problem related to https://github.com/checkpoint-restore/criu/pull/1643

Note: test mntns_ghost01 detects the problem fixed by 4-th patch when error handling  (1-2 patches) is fixed.

### CODE REVIEW

File: criu/files-reg.c
Comment (nviennot) [inline criu/files-reg.c:1797]: Yes, it's better to be safe

File: criu/files-reg.c
Comment (nviennot) [inline criu/files-reg.c:2150]: `pr_perror("Failed to unlink the remap file %d:%s", mntns_root, rfi->path);`

File: criu/mount.c
Comment (nviennot) [inline criu/mount.c:2740]: Shouldn't we free `remounted_rw` in `mnt_entry_free()` ?

File: criu/files-reg.c
Comment (Snorch) [reply criu/files-reg.c:2150]: - mntns_root is a ROOT_FD_OFF service fd number, I believe printing it has not much value
- the pr_debug above already prints mnt_id+path so we can skip it in pr_perror

File: criu/mount.c
Comment (Snorch) [reply criu/mount.c:2740]: No we can not, with shmalloc we can only free last allocation (see shfree_last) so in general we can't free shmalloc'ed data.

File: criu/files-reg.c
Comment (Snorch) [reply criu/files-reg.c:1797]: Something like https://github.com/Snorch/criu/commit/98f0445f884e3b2317d9efc74a77fd0a83bf48c9 can be done, but I'm not sure that we urgently need this because we've never seen such a race and all uses of path seem local not shared. But for future we should have this in mind.

Comment (nviennot) [reply criu/files-reg.c:1797]: Even if it's not racy, i would do it.

It's very surprising that a function that unlinks things actually modifies an argument. It makes the code more difficult to reason about, and thus more difficult to maintain.

Correctness is more important that saving a few cpu cycles

File: criu/files-reg.c
Comment (nviennot) [reply criu/files-reg.c:2150]: Log messages can get interleaved due to parallelism. It could br useful to have a more precise error message.

File: criu/mount.c
Comment (nviennot) [reply criu/mount.c:2740]: Oof okay I see. I thought from the name that "last" meant that the user should do reference counting or something.
shfree_last() shouldn't take an argument. I think I wouldn't have been confused if it was just `shfree_last()`

File: criu/files-reg.c
Comment (Snorch) [reply criu/files-reg.c:2150]: It's very common practice in criu that not only the error itself should be considered when debugging but also previous log messages. I get used to this practice probably too much =) So I would've just grep logs by pid to neutralize interleaved messages. But OK I can just make error message printing similar thing to pr_debug.

File: criu/files-reg.c
Comment (Snorch) [reply criu/files-reg.c]: > Correctness is more important that saving a few cpu cycles

Excess correctness checks are not always welcome =) One example that I can remember is --check-mounts patch (https://github.com/checkpoint-restore/criu/pull/660#issuecomment-548736668).

But anyway I tried to fix this comment. PR become a bit more complex. I'm OK to revert it to the original form in case @checkpoint-restore/maintainers are against this "strdup and staff" change.

**Comment by Snorch** [comment]
Oops, stupid compilation errors fixed.

**Comment by Snorch** [comment]
```
 b'(00.034649)      1: uffd: Lazy pages are not available: Operation not permitted'
b'(00.034651)   7809: uffd: Sending PID -7809'
b'(00.034654)      1: Error (criu/uffd.c:314): uffd: Unable to open an userfaultfd descriptor: Operation not permitted'
```
Don't think these errors are related to the patch...

**Comment by Snorch** [comment]
I've reverted debugging of https://github.com/checkpoint-restore/criu/issues/1633 as I've found the root cause.

File: criu/files-reg.c
Comment (nviennot) [reply criu/files-reg.c]: > Excess correctness checks are not always welcome
I agree. I don't think using defensive programming is useful in most cases.

Here, we are calling `rm_parent_dirs(mntns_root, rfi->path, level)`. I'm willing to bet that most programmers will say that It's unexpected that this call modifies the rfi->path string for two reasons 1) semantically, there's no reason that unlink()/rm_dirs() modifies its arguments 2) we are modifying the content of a struct. There's no reason to think that this string is temporary or won't be reused.

This can lead to future bugs. I stand strong on this stance.

For implementation details, it can be easier to reason about it by putting the dup logic outside of the function:

```c
int __rm_parent_dirs(..., char *path, ...) {
    ...
}

int rm_parent_dirs(..., const char *path, ...) {
    char *dup_path = strdup(path);
    if (!dup_path)
        return -1;
    int ret = __rm_parent_dirs(..., dup_path, ...);
    free(dup_path);
    return ret;
}
```

File: criu/files-reg.c
Comment (nviennot) [reply criu/files-reg.c]: Fair, I don't feel strongly about this one, your call

Comment (Snorch) [reply criu/files-reg.c]: So let's revert it to my variant then.

File: criu/files-reg.c
Comment (Snorch) [reply criu/files-reg.c]: > Here, we are calling rm_parent_dirs(mntns_root, rfi->path, level). I'm willing to bet that most programmers will say that It's unexpected that this call modifies the rfi->path string for two reasons 1) semantically, there's no reason that unlink()/rm_dirs() modifies its arguments 2) we are modifying the content of a struct. There's no reason to think that this string is temporary or won't be reused.

All modifications of rm_parent_dirs are invisible to the caller, it only temporary inserts '\0' in place of '\' and reverts everything back later.

> This can lead to future bugs. I stand strong on this stance.

That's why I initially wanted to add FIXME comment, so that it would be easier to track such bugs in case we would have them in future. And now I tend to think that fixing rm_parent_dirs argument modification can be a matter of another PR. This pr goal is to fix remount writable code, and fix error handling so that the problem with remount writable code would be visible (test mntn_ghost01 starts to fail).

> For implementation details, it can be easier to reason about it by putting the dup logic outside of the function:

The Idea looks good, but as we need to intrusively change rm_parent_dirs making it not replace '\0' back to '\' as this code becomes excess, so I don't think we need __rm_parent_dirs ...

Let's be lazy an leave strdup like I did it, and wait for some more review.

Comment (nviennot) [reply criu/files-reg.c]: > All modifications of rm_parent_dirs are invisible to the caller, it only temporary inserts '\0' in place of '\' and reverts everything back later.

Ohh okay, well in that case, it's fine. Nevermind.

**Comment by Snorch** [comment]
No comments on it for a long time...
I consider xstrdup and inplace_dirname changes too heavy to bee easily reviewed, and also not actually needed because it does not fix actual errors =)

So let's revert everything to the original state, other bonus from this is that I would have less pain porting other mount-v2 patches on top of it. Also let's remove FIXME completely as it prove itself confusing.

All current review comments are resolved, would like to have more review, to be able to finally merge this.

**Comment by rst0git** [comment]
>I believe it's a bug in centos8 kernel that IPPROTO_MPTCP_KERN is visible to userspace, in centos9 they don't have this problem.

`IPPROTO_MPTCP_KERN` has been added in RHEL with this commit:
```
[net] add and use MPTCP_PROTO_KERN


Message-id: <20200515101824.19445-21-fwestpha@redhat.com>
Patchwork-id: 309886
Patchwork-instance: patchwork
O-Subject: [RHEL8.3 net 20/38] net: add and use MPTCP_PROTO_KERN
Bugzilla: 1835796
RH-Acked-by: Paolo Abeni's avatarPaolo Abeni <pabeni@redhat.com>
RH-Acked-by: Marcelo Ricardo Leitner's avatarMarcelo Leitner <mleitner@redhat.com>
RH-Acked-by: Jarod Wilson's avatarJarod Wilson <jarod@redhat.com>
RH-Acked-by: Tony Camuso's avatarTony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1835796


Upstream Status: RHEL ONLY

Upstream kernel simply increased IPPROTO_MAX, which we can't
do due to KABI implications (the define is used e.g. for array sizing).

This patch adds a kernel-internal MPTCP_PROTO_KERN and
maps UAPI IPPROTO_MPTCP to it, i.e. if userspace asks for
protocol 262 (IPPROTO_MPTCP), the kernel changes that to "66" under
the hood so we maintain UAPI with upstream kernel.
```

**Comment by adrianreber** [comment]
> To @adrianreber : can we just disable socket-raw test on centos 8?

Works for me.

**Comment by Snorch** [comment]
> > To @adrianreber : can we just disable socket-raw test on centos 8?
> 
> Works for me.

I'm not too experienced with our ci, so advice on how to properly pass -x socket-raw would be appreciated.
https://github.com/checkpoint-restore/criu/pull/1658

**Review by nviennot — APPROVED** [review]
Great! Thank you @Snorch !

File: criu/files-reg.c
Comment (rst0git) [inline criu/files-reg.c:2038]: The usage of `tmp` and `ret` is slightly confusing here because we have `return tmp;` and `ret` is used as temporary variable.

File: criu/files-reg.c
Comment (nviennot) [inline criu/files-reg.c:1800]: This if statement is not useful (the while loop condition is sufficient), but I don't think it matters.

File: criu/files-reg.c
Comment (Snorch) [reply criu/files-reg.c:2038]: I believe that a better name for "tmp" is "fd" or "tmp_fd", also I think "ret" is in a proper place here, though it may be a bit confusing.

So I don't want to touch it as:
- "tmp" was named "tmp" for a long time
- patch goal is a bit unrelated - it is to fix error handling so that hidden errors became finally visible, not to fix variable naming
- Using "ret" as a temporary variable name to check called function ret code is a common practice, ret is not always a return value and it's ok (more over we tend to use "exit_code" name for return values in multiple places)

Comment (rst0git) [reply criu/files-reg.c:2038]: I agree with you but what do you think about renaming "tmp" in a separate patch to improve readability?

File: criu/files-reg.c
Comment (Snorch) [reply criu/files-reg.c]: Thanks! Let's remove it.

**Review by adrianreber — APPROVED** [review]
Sounds good. CI is happy.

**Review by rst0git — APPROVED** [review]

File: criu/files-reg.c
Comment (avagin) [inline criu/files-reg.c:2138]: pls check in the test that a test mount is read-only after restore.

File: criu/files-reg.c
Comment (avagin) [inline criu/files-reg.c:2161]: nit: I would rename it to err_with_unlock or something like this to make it obvious that it unlocks remap_open_lock

File: criu/files-reg.c
Comment (Snorch) [reply criu/files-reg.c:2138]: done:
zdtm: add ro-mount check after c/r to mntns_ghost01

File: criu/files-reg.c
Comment (Snorch) [reply criu/files-reg.c:2161]: Please don't:

1) only unlocks for remap case
2) extra 9 lines change which does not directly related to commit purpose

**Comment by Snorch** [comment]
Fails for Vagrant Fedora based test (no VDSO) check same as here:
https://github.com/checkpoint-restore/criu/pull/1664

So looks unrelated to this PR.

**Comment by adrianreber** [comment]
> Fails for Vagrant Fedora based test (no VDSO) check same as here: #1664
> 
> So looks unrelated to this PR.

Yes, see #1669

**Pull request merged**
